sources/academy/webscraping/scraping_basics_python/05_parsing_html.md

```diff
@@ -0,0 +1,211 @@
+
+---
+
+From lessons about browser DevTools we know that the HTML tags representing individual products have a `class` attribute which, among other values, contains `product-item`.
+
+![Products have the ‘product-item’ class](./images/product-item.png)
+
+As a first step, let's try counting how many products are on the listing page.
+
+## Treating HTML as a string
+
+Currently, the entire HTML is available in our program as a string. Our program can print it to the screen or save it to a file, but not much more. If it's a string, could we use Python string operations to count the products? Each Python string has `.count()`, a [method for counting substrings](https://docs.python.org/3/library/stdtypes.html#str.count).
+
+After manually inspecting the page in browser DevTools we can see that all product cards have the following structure:
+
+```html
+<div class="product-item product-item--vertical ...">
+  <a href="/products/..." class="product-item__image-wrapper">
+    ...
+  </a>
+  <div class="product-item__info">
+    ...
+  </div>
+</div>
+```
+
+At first sight, counting `product-item` occurrences wouldn't match only products, but also `product-item__image-wrapper`. Hmm.
+
+We could try looking for `<div class="product-item`, a substring which represents the enitre beginning of each product tag, but that would also count `<div class="product-item__info`! We'll need to add a space after the class name to avoid matching those. Replace your program with the following code:
+
+```py
+import httpx
+
+url = "https://warehouse-theme-metal.myshopify.com/collections/sales"
+response = httpx.get(url)
+response.raise_for_status()
+
+html_code = response.text
+count = html_code.count('<div class="product-item ')
+print(count)
+```
+
+Note that because the substring contains a double quote character, we need single quotes as string boundaries.
+
+:::info Handling errors
+
+To have the code examples more concise, we're omitting error handling for now. Keeping `response.raise_for_status()` ensures that your program at least visibly crashes and prints what happened in case there's an error.
+
+:::
+
+Our scraper prints 24, which is in line with the text **Showing 1–24 of 50 products** above the product listing. Phew, figuring this out was quite tedious!
+
+```text
+$ python main.py
+24
+```
+
+<!-- TODO image -->
+
+While possible, we can see that processing HTML with [standard string methods](https://docs.python.org/3/library/stdtypes.html#string-methods) is difficult and fragile. Imagine we wouldn't be just counting, but trying to get the titles and prices.
+
+In fact HTML can be so complex that even [regular expressions](https://docs.python.org/3/library/re.html) aren't able to process it reliably. To work with HTML we need a robust tool dedicated for the task.
+
+:::tip Why regex can't parse HTML
```

Comment:
nit: not sure if `tip` is the best category used here, maybe `info`?

Reply:
Addressed in c777e268